src/StreamJsonRpc.Tests/MessagePackFormatter.cs

```diff
@@ -119,5 +171,95 @@ public void Serialize(IBufferWriter<byte> contentBuffer, JsonRpcMessage message)
 
             return result;
         }
+
+        private class StandardPlusIProgressOfTResolver : IFormatterResolver
+        {
+            private readonly MessagePackFormatter formatter;
+
+            public StandardPlusIProgressOfTResolver(MessagePackFormatter formatter)
+            {
+                this.formatter = formatter;
+            }
+
+            public IMessagePackFormatter<T> GetFormatter<T>()
+            {
+                return new FormatterCache<T>(this.formatter).Formatter;
+            }
+
+            private class FormatterCache<T>
+            {
+                public readonly IMessagePackFormatter<T> Formatter;
+
+                public FormatterCache(MessagePackFormatter formatter)
```

Comment:
Don't use the caching strategy that MessagePack documents or uses elsewhere, because your formatter is absolutely stateful and correlated to a particular JsonRpc instance. You should cache the formatters, but you should do so locally and not trying to use the .NET type system as the MessagePack library (perhaps inappropriately) does.